src/app/security/maturity-model/utils/useSectionHashObserver.ts

```diff
@@ -0,0 +1,84 @@
+'use client'
+
+import { useEffect, useState } from 'react'
+
+import { useRouter, usePathname } from 'next/navigation'
+
+import type { Route } from 'next'
+
+export const HASH_SIGN = '#'
+
+export type SectionHash = `${typeof HASH_SIGN}${string}`
+type NewHashEvent = HashChangeEvent | CustomEvent<SectionHash>
+
+export function useSectionHashObserver() {
+  const pathname = usePathname()
+  const router = useRouter()
+
+  const [hash, setHash] = useState(getHashFromWindow)
+
+  function updateSectionHash(sectionId: string) {
+    const sectionHash: SectionHash = `${HASH_SIGN}${sectionId}`
+    const pathnameWithNewSectionHash = pathname + sectionHash
+
+    router.replace(pathnameWithNewSectionHash as Route, { scroll: false })
+
+    if (windowIsDefined()) {
+      window.dispatchEvent(
+        new CustomEvent('hashchange', { detail: sectionHash }),
+      )
+    }
+  }
+
+  function clearSectionHash() {
+    router.replace(pathname as Route, { scroll: false })
+
+    if (windowIsDefined()) {
+      window.dispatchEvent(new CustomEvent('hashchange', { detail: HASH_SIGN }))
+    }
+  }
+
+  function onHashChange(event: NewHashEvent) {
+    const newHash = getHashFromEvent(event)
+    setHash(newHash)
+  }
+
+  useEffect(() => {
+    if (windowIsDefined()) {
+      window.addEventListener('hashchange', onHashChange)
+      return () => window.removeEventListener('hashchange', onHashChange)
+    }
+  }, [pathname])
```

Comment:
Is adding the listener only once not enough?

```suggestion
  useEffect(() => {
    if (windowIsDefined()) {
      window.addEventListener('hashchange', onHashChange)
      return () => window.removeEventListener('hashchange', onHashChange)
    }
  }, [])
```